### ISSUE TITLE
DDeflect wishlist

### ISSUE THREAD

**Issue by nosmo** [equalitie/bundler#1]
* A configuration file - It can be JSON/YAML/INI, something readable that is easily understood cross-language. My preference is for YAML or JSON. 
* Daemonisation - ability to run as a service. Not a high priority while dev is in progress. 
* Privilege dropping - process.setgid/process.setuid
* Polite logging - ability to verbosely log to console or to actionably log to syslog. Explanatory logging once a failure is detected (currently all we have is Abort: fail from PhantomJS :poop:)
* Parallelised fetching
* Encoding awareness - probably not a major issue with Node but we have encountered some issues with coercion of strings breaking encodings. I assume there is a Javascript replacement for Python’s `chardet` library. 
* Remaps - the ability to read a data structure that will say www.site.com is actually at 10.0.0.1, and then remap the origin based on the Host header of the incoming request (in addition to other header manipulations/passing elsewhere ) 
* Proxy support - the ability to use a proxy server for all requests to the origin site
* Header-passing support - incoming requests will have headers that will need to be passed on to the origin (cookies, other headers) and some that will not need to be passed on (Cache-Control etc). A white/blacklist system for this would be great. 
* Redirect handling - Initially support for returning redirects, later support for all HTTP status codes
* Polite behaviours - passing a proper Via header, setting a user agent

**Comment by redwire** [comment]
This looks like a pretty good list.  It should be interesting to implement all of this.  I'd like to ask a few questions on a point-by-point basis so that I understand everything correctly.

1. Configuration file - Could you perhaps give me a list of the kinds of things you might want to be able to configure? I can certainly try to integrate configuration options for a number of the other points, but if there is anything else you had in mind, please do mention them!
2. Daemonisation - Would this apply to a web server? As it stands, bundler is a library so I would like to be a little more clear on what service would be running as a daemon.
3. Privilege dropping - Is it the metaphorical bundler server that needs to have privileges reduced? If so, I'm not sure that will fit into the bundler library by itself.
4. Encoding awareness - Is it the case, here, that we may not want to encode all resources in base64? Could you give me an idea (or even a complete list) of what resources might need to have different encodings in their Data-URI-Scheme?
5. Remaps - I'm really not sure what you mean here at all other than that there might be some IP address lookup going on. Are you referring to doing Referrer header spoofing?
6. Proxy server - I feel like a proxy server utilizing bundler might best be its own project, or else a rather small application of it

Thanks!

**Comment by redwire** [comment]
Just for my own convenience, I am going to keep a check-list here of items I have completed.

* [x] A configuration file
* [  ] Daemonisation
* [  ] Privilege dropping
* [  ] Polite logging
* [  ] Parallelized fetching
* [  ] Encoding awareness
* [  ] Remaps
* [  ] Proxy support
* [  ] Header-passing support
* [  ] Redirect handling
* [  ] Polite behaviors

**Comment by nosmo** [comment]
It seems I may have misunderstood what Bundler the library is as opposed to a service that wraps Bundler as a webserver or something else of that nature, my mistake. For this reason you can ignore: 

Daemonisation
Privilege dropping

And *probably* Configuration file - the only configuration options I can think of passing would be HTTP proxy details. 

When I referred to "Proxy support" I actually meant that Bundler should support sending its GETs through an external proxy server. This means that if we need it we can deploy a caching proxy in order to speed up bundling time on slow origin servers.

**Comment by nosmo** [comment]
For the remap support (which would be better labelled something like reverse proxying), I think it might be easier if I simply give an example: 

When DDeflect is protecting site.com, a server (server.deflect.ca) running Bundler will appear in DNS records for site.com instead of the origin server for site.com (let's call it origin.site.com). When server.deflect.ca sees a HTTP request for "Host: site.com" it will use an internal lookup to realise that the resources for site.com are actually on server.site.com. Bundler will request resources from server.site.com setting the Host header to site.com, keeping server.site.com hidden while returning the results to the user for site.com as normal.

**Comment by redwire** [comment]
Thanks for these pieces of clarification.  There's a lot there that I can act on.
I started work on pretty much rearchitecting the bundler library yesterday so that it will support the insertion of hook functions:

1. Before the original request
2. Before making a request for a resource
3. After collecting the {source: <resource url>, destination: <data-uri>} pairs for filtering/analysis

If you could take a brief moment to look over some of the [refactoring I did in the first relevant commit](https://github.com/equalitie/bundler/commit/66066f550db93a1d1b9fa0e5c49855dd2ee6a3ff) and let me know if you think this approach will allow for the kind of functionality you're looking for, I'd appreciate that.

**Comment by redwire** [comment]
@nosmo Could you give me an idea of where the data structure you are talking about for doing lookups for resources during remapping/reverse proxying will come from? How will any bundler handler code I write actually interact with it?

**Comment by redwire** [comment]
I have found in my readings about parallelizing code run on Node.js, that it 
is impractical to expect to be able to reasonably create a threaded request
maker or else use separate processes.  Unless it is in fact another part of
the system slowing things down (which we may want to investigate), it is likely
that resource requests are not going to get much faster in Javascript.

**Comment by redwire** [comment]
@nosmo Could you comment on whether what I  pushed in 
https://github.com/equalitie/bundler/commit/83be42971999e1061d22a53065536289ee5b0d05
is roughly what you have in mind for remapping?

**Comment by nosmo** [comment]
Sorry for the delay on responding to this - the lookup structure is a simple map in my mind, where hostname maps to IP address. Looking at 83be42971999e1061d22a53065536289ee5b0d05 it seems that that fully satisfies the requirements - assuming that remapper would be something like {"fulltimeinter.net": 92.51.245.61} - am I correct in this assumption?

**Comment by redwire** [comment]
That is about what I expected it to be. The map would have values that are hostnames of sites on which resources are hosted and the values would be the (string) address to change the hostname to.

**Comment by nosmo** [comment]
Could you include an example psconfig.json please? I've got the server running with a dummy config but I'd like to be sure.

**Comment by nosmo** [comment]
I assume I have things set up wrong, but could you direct me on what to do here? 

In proxyserver.js I've defined ```var remapper = {"equalit.ie": "92.51.245.61"};```

When I run ```curl -H "Host: equalit.ie" 127.0.0.1:9008/```

I get
```
Proxy server listening on 127.0.0.1:9008
debug: in Bundler.send/async.reduce, memo = url=/
debug: in Bundler.send/async.reduce, memo = url=/, User-Agent=undefined, Accept=undefined
debug: in Bundler.send/async.reduce, memo = url=/, User-Agent=undefined, Accept=undefined
debug: in send, options = %j url=/, User-Agent=undefined, Accept=undefined, followRedirect=undefined, followAllRedirects=undefined, maxRedirects=undefined
error: Error making request to /; Error: Invalid URI "/"
Failed to create bundle for /
Error: Invalid URI "/"
```

**Comment by redwire** [comment]
@nosmo 

I have uploaded a [config file for the proxy server](https://github.com/equalitie/bundler/blob/master/applications/psconfig.json) that you are welcome to reference.  I have also provided default configuration settings in https://github.com/equalitie/bundler/commit/c7ab5935bcc237fbf10f808b4d63ead2eda71c96 that will be extended by the config options so that you do not have to provide all of them.
Note that the `cloneHeaders` field uses the names for headers that are provided by [http.IncomingMessage.headers](http://nodejs.org/api/http.html#http_message_headers).

As for running the proxy server, I have been testing it by configuring my browser to use `localhost:9008` as a proxy server.  That is, a request for a document (such as "https://www.equalit.ie") is made to the browser like normal and proxied through the bundler server.

Related: I've tested equalit.ie and deflect.ca, and the latter loads fine it seems other than the icon (which is a CSS background-image- i.e. a specific case of a handler yet to be written). While equalit.ie isn't rendering very well, I've noticed the problem has to do with the CSS documents being identified as not having the text/css mimetype, and this should be a minor fix.

**Comment by nosmo** [comment]
Consistently getting the following when testing with any URL, I'll look into what's causing it in a few minutes though: 
```
error: Error making request to undefined; Error: options.uri is a required argument
Failed to create bundle for /
Error: options.uri is a required argument
```

**Comment by redwire** [comment]
@nosmo See https://github.com/equalitie/bundler/commit/e89d9e25e815b7eb7346cf46012c25cb3dcb930d

**Comment by nosmo** [comment]
Ah, I think I understand. I am getting bundles back at least. I don't really know node and haven't been reading the code so I'm only testing this for functional stuff - please point out when I need to do stuff like this. 

Do you know how much time would be required to add reverse proxy style functionality? Query another resource based on the Host header and supplied path, and so on.

**Comment by redwire** [comment]
Is that not the remapping thing? I thought we agreed that was essentially done.

**Comment by nosmo** [comment]
Not quite - it means that if Bundler is running on localhost I can do ```curl -H "Host: equalit.ie" localhost:9008/test``` and get equalit.ie/test back. If this is a hassle to implement we can continue with the current implementation just fine

**Comment by redwire** [comment]
That'd just be something you'd always do just for the original request? I think that could be easy to do.

**Comment by redwire** [comment]
Updated comment

**Comment by nosmo** [comment]
Hmm... I've had a bit of a think about this. I'd be fully happy with option 1 if there weren't concerns about the query string issue. We can pass additional parameters and/or headers as we need them for additional IPC. 

However, I'm not quite sure I understand your concern with option 2 - in the setup I am imagining, all that would be happening is a single GET for site.com arrives at the server, asking for a bundle of site.com. The Bundler proxy itself wouldn't be the endpoint for visits to site.com, there'll be a manager and a caching proxy in front of them so generally (but not always sadly) requests coming from the client will be for page content rather than resources.

**Comment by redwire** [comment]
When I was testing the proxy server with configuration 2, I had a lot of output coming in from attempts at bundling all the resources that had to be fetched but weren't being bundled.
So if I configured the bundler to only bundle CSS files, then when the page loaded all the requests for images (for example) would be made by my browser and go to the proxy server,  which would attempt to bundle them.  Obviously it can't bundle individual resources normally so it was just creating a lot of noise that made it hard to debug.

**Comment by nosmo** [comment]
Ah, I see. Could you outline the pitfalls you see for approach number 1? I can imagine that url paths and so on will need to be escaped, which will lead to a double escaping. This in and of itself isn't the end of the world, as we'll just unescape one level and continue onwards. I anticipate passing headers directly from the original request at some point, they will be unaffected if we pass them on as normal.

**Comment by redwire** [comment]
Well, as I wrote above:

* It's not very configurable- I could not really get the Host information I wanted out of the requests and URLs I tried to fabricate with curl using this approach
* It's probably not very practical for deployment purposes
* Links on web pages will need to be specially formatted to be sent through the query string

These might be non-issues, however.

When I used approach 1 and tried to get the Host header from the request object in the server, I would always get `localhost:9008/` even when setting the header with CURL. My fear that it might not be practical in deployment comes from the fact that users of the proxy server would have to craft these specially formatted URLs to fetch a particular document. While that's not hard, it might be a bit much to ask if the proxy is meant to run as a user-facing service at any point.
